engine/session/terminal.go

```
@@ -0,0 +1,157 @@
+	rootCtx context.Context,
+	withTerminal WithTerminalFunc,
+) TerminalAttachable {
+	if withTerminal == nil {
+		withTerminal = func(fn func(stdin io.Reader, stdout, stderr io.Writer) error) error {
+			return fn(os.Stdin, os.Stdout, os.Stderr)
+		}
+	}
+
+	return TerminalAttachable{
+		rootCtx:      rootCtx,
+		withTerminal: withTerminal,
+	}
+}
+
+func (s TerminalAttachable) Register(srv *grpc.Server) {
+	RegisterTerminalServer(srv, s)
+}
+
+func (s TerminalAttachable) Session(srv Terminal_SessionServer) error {
+	return s.withTerminal(func(stdin io.Reader, stdout, stderr io.Writer) error {
+		return s.session(srv, stdin, stdout, stderr)
+	})
+}
+
+func (s TerminalAttachable) session(srv Terminal_SessionServer, stdin io.Reader, stdout io.Writer, stderr io.Writer) error {
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+
+	if err := s.sendSize(srv, stdout); err != nil {
+		return fmt.Errorf("sending initial size: %w", err)
+	}
+	go s.listenForResize(ctx, srv, stdout)
+	go s.forwardStdin(ctx, srv, stdin)
+
+	for {
+		req, err := srv.Recv()
+		if err != nil {
+			if errors.Is(err, context.Canceled) || grpcerrors.Code(err) == codes.Canceled {
+				// canceled
+				return nil
+			}
+
+			if errors.Is(err, io.EOF) {
+				// stopped
+				return nil
+			}
+
+			if grpcerrors.Code(err) == codes.Unavailable {
+				// client disconnected (i.e. quitting Dagger out)
+				return nil
+			}
+
+			return fmt.Errorf("error reading terminal: %w", err)
+		}
+		switch msg := req.GetMsg().(type) {
+		case *SessionRequest_Stdout:
+			stdout.Write(msg.Stdout)
+		case *SessionRequest_Stderr:
+			stderr.Write(msg.Stderr)
+		case *SessionRequest_Exit:
+			fmt.Fprintf(stderr, "exit %d\n", msg.Exit)
+			return nil
+		}
+	}
+}
+
+func (s TerminalAttachable) sendSize(srv Terminal_SessionServer, stdout io.Writer) error {
+	f, ok := stdout.(*os.File)
+	if !ok || !isatty.IsTerminal(f.Fd()) {
+		slog.Debug("stdin is not a terminal; cannot get terminal size")
+		return nil
```

seems like it should just return an err in this case I think?